pkg/v1/tarball/layer.go

```diff
@@ -281,3 +286,21 @@ func computeDiffID(opener Opener) (v1.Hash, error) {
 	digest, _, err := v1.SHA256(rc)
 	return digest, err
 }
+
+// reopener returns an Opener that returns rc on the first call and opener()
+// on subsequent calls. This is used in conjunction with gzip.Is to avoid
+// calling an opener just to read two bytes.
+func reopener(rc io.ReadCloser, opener Opener) Opener {
+	mu := sync.Mutex{}
+	consumed := false
+	return func() (io.ReadCloser, error) {
+		mu.Lock()
+		defer mu.Unlock()
+
+		if !consumed {
+			consumed = true
+			return rc, nil
+		}
```

Comment:
I'd be there's a way to do this more efficiently for reopens with `sync.Once` wrt synchronization.

Reply:
Something like this?

```go
func reopener(rc io.ReadCloser, opener Opener) Opener {
        var once sync.Once
        return func() (io.ReadCloser, error) {
                open := opener
                once.Do(func() {
                        open = func() (io.ReadCloser, error) {
                                return rc, nil
                        }
                })
                return open()
        }
}
```

I don't think that's actually safe and benchmarking reveals no substantial speedup either way. Do you have something else in mind?

Reply:
> I don't think that's actually safe

Why not?

Reply:
I guess the more likely path is `if consumed` so maybe it makes sense to move the `if !consumed` path into a separate function... I guess we could copy over the `sync.Once` implementation?

Reply:
Here's `sync.Once` inlined, without the "slow path" outlined:

```go
func reopener(rc io.ReadCloser, opener Opener) Opener {
        var (
                mu       sync.Mutex
                consumed uint32
        )
        return func() (io.ReadCloser, error) {
                if atomic.LoadUint32(&consumed) == 0 {
                        mu.Lock()
                        defer mu.Unlock()

                        if consumed == 0 {
                                defer atomic.StoreUint32(&consumed, 1)
                                return rc, nil
                        }
                }

                return opener()
        }
}
```

Reply:
> Why not?

Eh... I was thinking because the `return open()` happens outside of the sync.Once mutex, we'd have a race between goroutines, but thinking about it, we're mutating the `open` on the stack of whichever goroutine hit that first, so it doesn't matter at all? 🤷

Reply:
Pushed with this impl: https://github.com/google/go-containerregistry/pull/885#discussion_r548260635

Reply:
I would use: https://godoc.org/sync/atomic#CompareAndSwapInt32

Reply:
Done